src/osmdata-data_frame.cpp

```diff
@@ -0,0 +1,267 @@
+ *  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+ *  details.
+ *
+ *  You should have received a copy of the GNU General Public License along with
+ *  osmdata.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ *  Author:     Mark Padgham 
+ *  E-Mail:     mark.padgham@email.com 
+ *
+ *  Description:    Modified version of 'osmdata-sf' to extract OSM data from an
+ *                  object of class XmlData and return it in Rcpp::List format,
+ *                  but in this case ignoring the actual geometric data. The
+ *                  returned object contains key-value data only, for return
+ *                  from the R function 'osmdata_data_frame'.
+ *
+ *  Limitations:
+ *
+ *  Dependencies:       none (rapidXML header included in osmdata)
+ *
+ *  Compiler Options:   -std=c++11
+ ***************************************************************************/
+
+#include "osmdata.h"
+
+#include <Rcpp.h>
+#include <list>
+#include <string>
+
+// Note: This code uses explicit index counters within most loops which use Rcpp
+// objects, because these otherwise require a 
+// static_cast <size_t> (std::distance (...)). This operation copies each
+// instance and can slow the loops down by several orders of magnitude!
+
+/************************************************************************
+ ************************************************************************
+ **                                                                    **
+ **          1. PRIMARY FUNCTIONS TO TRACE WAYS AND RELATIONS          **
+ **                                                                    **
+ ************************************************************************
+ ************************************************************************/
+
+
+//' get_osm_relations
+//'
+//' Return a dual Rcpp::DataFrame containing all OSM relations.
+//'
+//' @param rels Pointer to the vector of Relation objects
+//' @param unique_vals Pointer to a UniqueVals object containing std::sets of all
+//'       unique IDs and keys for each kind of OSM object (nodes, ways, rels).
+//'
+//' @return A dual Rcpp::DataFrame with the tags and metadata of the relations.
+//' 
+//' @noRd 
+Rcpp::DataFrame osm_df::get_osm_relations (const Relations &rels, 
+        const UniqueVals &unique_vals)
+{
+
+    const unsigned int nmp = rels.size ();
+
+    size_t ncol = unique_vals.k_rel.size ();
+    std::vector <std::string> rel_ids; 
+    rel_ids.reserve (nmp);
+
+    Rcpp::CharacterMatrix kv_mat (Rcpp::Dimension (nmp, ncol));
+    unsigned int count = 0;
+
+    for (auto itr = rels.begin (); itr != rels.end (); ++itr)
+    {
+        if (count % 1000 == 0)
+            Rcpp::checkUserInterrupt ();
+
+        rel_ids.push_back (std::to_string (itr->id));
+
+        osm_convert::get_value_mat_rel (itr, unique_vals, kv_mat, count++);
+    }
+
+    Rcpp::DataFrame kv_df;
+    if (nmp > 0)
+    {
+        kv_mat.attr ("names") = unique_vals.k_rel;
```

Comment:
```suggestion
```
`names` attribute is the same as the second dimension of the `dimnames` assigned in the following line